proposals/nnnn-inline-spirv-and-decorated-types.md

```diff
@@ -0,0 +1,132 @@
+<!-- {% raw %} -->
+
+# Target Extension Types for Inline SPIR-V and Decorated Types
+
+* Proposal: [NNNN](NNNN-filename.md)
+* Author(s): [Cassandra Beckley](https://github.com/cassiebeckley)
+* Status: **Design In Progress**
+* PRs: [#125316](https://github.com/llvm/llvm-project/pull/125316)
+
+## Introduction
+
+In this proposal, we define the `spirv.Type` and `spirv.DecoratedType`
+target extension types for the SPIR-V backend.
+
+## Motivation
+
+We would like to implement `SpirvType` and `SpirvOpaqueType` from [Inline
+SPIR-V (HLSL proposal 0011)](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0011-inline-spirv.md#types)
+in the SPIR-V backend. These allow users to create arbitrary `OpType`
+instructions, and use them like any other HLSL type. Using these types, a
+vendor creating a new extension can expose it to users by creating a header
+file without needing to modify the compiler.
+
+Additionally, we need a way to represent types with SPIR-V decorations in LLVM
+IR.
+
+## Proposed solution
+
+### SpirvType
+
+To represent `vk::SpirvType` and `vk::SpirvOpaqueType` in LLVM IR, we will add
+three new target extension types:
+
+| Type                     | HasZeroInit | CanBeGlobal | CanBeLocal |
+| ------------------------ | ----------- | ----------- | ---------- |
+| `spirv.Type`             | false       | true        | true       |
+| `spirv.IntegralConstant` | false       | false       | false      |
+| `spirv.Literal`          | false       | false       | false      |
+
+`IntegralConstant` and `Literal` are used to encode arguments to `Type`, and
+may not be used outside that context. They are necessary because target
+extension types must have all type arguments precede all integer arguments,
+whereas SPIR-V type instructions may have an arbitrary number of type,
+immediate literal, and constant id operands in any order.
+
+#### `spirv.Type`
+
+```
+target("spirv.Type", operands..., opcode, size, alignment)
+```
+
+`opcode` is an integer literal representing the opcode of the `OpType`
+instruction to be generated. `size` and `alignment` are integer literals
+representing the number of bytes a single value of the type occupies and the
+power of two that the value will be aligned to in memory. An opaque type can be
```

Comment:
nit: I don't think the spec requires the alignment to be a power of 2

Reply:
Hm, this value is specified as being a power of 2 in the [HLSL `SpirvType` proposal](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0011-inline-spirv.md). It makes implementation on the backend simpler as well, as we can just use an appropriately-sized int as the layout type. If we have specific requests from users maybe we can relax this requirement, but otherwise I think it would be fine to keep it.